custodial-copy-confirmer/src/main/scala/uk/gov/nationalarchives/confirmer/Main.scala

```diff
@@ -0,0 +1,79 @@
+import pureconfig.*
+import pureconfig.module.catseffect.syntax.*
+import uk.gov.nationalarchives.{DADynamoDBClient, DASQSClient}
+import io.circe.generic.auto.*
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider
+import software.amazon.awssdk.http.nio.netty.{NettyNioAsyncHttpClient, ProxyConfiguration}
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue
+import uk.gov.nationalarchives.DADynamoDBClient.DADynamoDbRequest
+
+import java.net.URI
+import scala.concurrent.duration.*
+import java.util.UUID
+
+object Main extends IOApp {
+
+  case class Config(sqsUrl: String, proxyUrl: URI, ocflRepoDir: String, ocflWorkDir: String) derives ConfigReader
+
+  extension (s: String)
+    private def toAttributeValue: AttributeValue = AttributeValue.builder.s(s).build
+
+  case class Message(dynamoKey: String, dynamoTable: String, ioRef: UUID, batchId: String) {
+    def primaryKey: Map[String, AttributeValue] =
+      Map("assetId" -> ioRef.toString.toAttributeValue, "batchId" -> batchId.toAttributeValue)
+
+    def toUpdate: Map[String, Option[AttributeValue]] =
+      Map(dynamoKey -> "true".toAttributeValue.some)
+  }
+
+
+  private def dynamoClient(proxyUrl: URI): DADynamoDBClient[IO] = {
+    val proxy = ProxyConfiguration
+      .builder()
+      .scheme(proxyUrl.getScheme)
+      .host(proxyUrl.getHost)
+      .port(proxyUrl.getPort)
+      .build
+    val dynamoDBClient: DynamoDbAsyncClient = DynamoDbAsyncClient
+      .builder()
+      .httpClient(NettyNioAsyncHttpClient.builder().proxyConfiguration(proxy).build())
+      .region(Region.EU_WEST_2)
+      .credentialsProvider(DefaultCredentialsProvider.create())
+      .build()
+
+    DADynamoDBClient[IO](dynamoDBClient)
+  }
+
+  override def run(args: List[String]): IO[ExitCode] =
+    for {
+      config <- ConfigSource.default.loadF[IO, Config]()
+      sqs = sqsClient[IO](config.proxyUrl.some)
+      dynamo = dynamoClient(config.proxyUrl)
+      _ <- (Stream.fixedRateStartImmediately[IO](10.seconds) >> runConfirmer(config, sqs, dynamo, Ocfl(config))).compile.drain
+    } yield ExitCode.Success
+
+
+  def runConfirmer(config: Config, sqsClient: DASQSClient[IO], dynamoClient: DADynamoDBClient[IO], ocfl: Ocfl): Stream[IO, Unit] = Stream.eval {
+    for {
+      logger <- Slf4jLogger.create[IO]
+      messages <- aggregateMessages[IO, Message](sqsClient, config.sqsUrl)
+      _ <- logger.info(s"Processing message refs ${messages.map(_.message.ioRef).mkString(",")}")
+      _ <- messages.parTraverse { sqsMessage =>
+        val message = sqsMessage.message
+        val request = DADynamoDbRequest(message.dynamoTable, message.primaryKey, message.toUpdate)
+        IO.whenA(ocfl.checkObjectExists(message.ioRef))(dynamoClient.updateAttributeValues(request).void) >>
```

Comment:
Just curious, since we're deleting it from the queue regardless of whether the object exists or not, what happens to the object that doesn't exist, does the Resender add it back into the queue so that the Confirmer can look at it later on?

Reply:
Ah, we're not supposed to delete from the queue if it doesn't exist :facepalm: Sorry, I'll change it.